openhands/server/session/session.py

```diff
@@ -52,41 +53,31 @@ def __init__(
         self.agent_session.event_stream.subscribe(
             EventStreamSubscriber.SERVER, self.on_event, self.sid
         )
-        self.config = config
+        # Copying this means that when we update variables they are not applied to the shared global configuration!
+        self.config = deepcopy(config)
         self.loop = asyncio.get_event_loop()
-        self.settings = None
 
     def close(self):
         self.is_alive = False
         self.agent_session.close()
 
-    async def initialize_agent(self, data: dict):
-        self.settings = data
+    async def initialize_agent(self, session_config: SessionConfig):
         self.agent_session.event_stream.add_event(
             AgentStateChangedObservation('', AgentState.LOADING),
             EventSource.ENVIRONMENT,
         )
         # Extract the agent-relevant arguments from the request
-        args = {key: value for key, value in data.get('args', {}).items()}
-        agent_cls = args.get(ConfigType.AGENT, self.config.default_agent)
-        self.config.security.confirmation_mode = args.get(
-            ConfigType.CONFIRMATION_MODE, self.config.security.confirmation_mode
-        )
-        self.config.security.security_analyzer = data.get('args', {}).get(
-            ConfigType.SECURITY_ANALYZER, self.config.security.security_analyzer
-        )
-        max_iterations = args.get(ConfigType.MAX_ITERATIONS, self.config.max_iterations)
+        agent_cls = session_config.agent or self.config.default_agent
+        self.config.security.confirmation_mode = self.config.security.confirmation_mode if session_config.confirmation_mode is None else session_config.confirmation_mode
+        self.config.security.security_analyzer = session_config.security_analyzer or self.config.security.security_analyzer
+        max_iterations = session_config.max_iterations or self.config.max_iterations
         # override default LLM config
+        
+
         default_llm_config = self.config.get_llm_config()
-        default_llm_config.model = args.get(
-            ConfigType.LLM_MODEL, default_llm_config.model
-        )
-        default_llm_config.api_key = args.get(
-            ConfigType.LLM_API_KEY, default_llm_config.api_key
-        )
-        default_llm_config.base_url = args.get(
-            ConfigType.LLM_BASE_URL, default_llm_config.base_url
-        )
+        default_llm_config.model = session_config.llm_model or default_llm_config.model
+        default_llm_config.api_key = session_config.llm_api_key or default_llm_config.api_key
+        default_llm_config.base_url = session_config.llm_base_url or default_llm_config.base_url
```

Comment:
seems like we're still mucking with the global defaults here

Reply:
It's the current behavior though. 🤔 What do you think it should do?

Reply:
Modifying global state with a single session would potentially create a ton of bugs and security issues, so I think the `deepcopy` we have above now makes this much safer. Will be much more important in a multi-Conversation world too (which we're working on now)

Reply:
Right, I'm not thinking of it as global state, but it kinda is. It's the fallback LLMConfig/etc instance one gets if they didn't make / retrieve a specific instance, right?

If what we need here is a an LLMConfig / SecurityConfig instance per sid, could we make an LLMConfig / SecurityConfig instance per sid?